File: js/src/blocklist.ts
import { BasehError } from "./errors.js";

export type BasehProfanityMode = "none" | "no-vowels" | "blocklist";

/** Spec 18. Optional profanity safety configuration. */
export interface BasehProfanity {
  mode: BasehProfanityMode;
  /** Replaces the default list when present (mode "blocklist" only). */
  words?: string[];
  /** Appended to the effective list (mode "blocklist" only). */
  extraWords?: string[];
}

/** Spec 18.2 default list. Deliberately small; applications extend it. */
export const DEFAULT_BLOCKLIST: readonly string[] = [
  "CRAP", "TWAT", "SHAG", "DAMN", "FCK", "FUC",
  "SHT", "CNT", "TWT", "DCK", "AZZ", "BCH"
];

const WORD = /^[A-Za-z]{2,32}$/;

function fail(reason: string): never {
  throw new BasehError("INVALID_PROFILE", `Invalid BaseH profile: ${reason}`, false);
}

/** Spec 18.2: replacement semantics, then augmentation, uppercased and deduplicated. */
export function effectiveBlocklist(profanity: BasehProfanity): string[] {
  const base = profanity.words ? [...profanity.words] : [...DEFAULT_BLOCKLIST];
  const list = [...base, ...(profanity.extraWords ?? [])];
  const out: string[] = [];
  for (const word of list) {
    if (typeof word !== "string" || !WORD.test(word)) {
      fail("blocklist entries must be 2 through 32 ASCII letters");
    }
    const upper = word.toUpperCase();
    if (!out.includes(upper)) out.push(upper);
  }
  return out;
}

/** Spec 18.1: vowels removed for no-vowels mode, applied after case normalization. */
export function stripVowels(alphabetNorm: string): string {
  return [...alphabetNorm].filter((c) => !"AEIOU".includes(c)).join("");
}

File: go/blocklist.go
package basehuman

import "strings"

// ProfanityMode selects the profanity safety strategy of spec 18.
type ProfanityMode string

// Profanity safety modes.
const (
	ProfanityNone      ProfanityMode = "none"
	ProfanityNoVowels  ProfanityMode = "no-vowels"
	ProfanityBlocklist ProfanityMode = "blocklist"
)

// Profanity is the optional profanity safety configuration of spec 18. A
// zero value means mode "none".
type Profanity struct {
	Mode ProfanityMode `json:"mode"`
	// Words replaces the default list when non-nil (blocklist mode only).
	Words []string `json:"words,omitempty"`
	// ExtraWords is appended to the effective list in either case.
	ExtraWords []string `json:"extraWords,omitempty"`
}

// DefaultBlocklist is the spec 18.2 default list. Deliberately small;
// applications extend it with Words or ExtraWords.
var DefaultBlocklist = []string{
	"CRAP", "TWAT", "SHAG", "DAMN", "FCK", "FUC",
	"SHT", "CNT", "TWT", "DCK", "AZZ", "BCH",
}

// isBlocklistWord reports whether s is 2 through 32 ASCII letters.
func isBlocklistWord(s string) bool {
	if len(s) < 2 || len(s) > 32 {
		return false
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if (c < 'A' || c > 'Z') && (c < 'a' || c > 'z') {
			return false
		}
	}
	return true
}

// effectiveBlocklist implements spec 18.2: replacement semantics, then
// augmentation, uppercased and deduplicated.
func effectiveBlocklist(p Profanity) ([]string, error) {
	base := p.Words
	if base == nil {
		base = DefaultBlocklist
	}
	list := make([]string, 0, len(base)+len(p.ExtraWords))
	list = append(list, base...)
	list = append(list, p.ExtraWords...)
	seen := make(map[string]bool, len(list))
	out := make([]string, 0, len(list))
	for _, word := range list {
		if !isBlocklistWord(word) {
			return nil, invalidProfile("blocklist entries must be 2 through 32 ASCII letters")
		}
		upper := strings.ToUpper(word)
		if !seen[upper] {
			seen[upper] = true
			out = append(out, upper)
		}
	}
	return out, nil
}

// stripVowels removes AEIOU for no-vowels mode, per spec 18.1. Applied to
// the case-normalized alphabets.
func stripVowels(alphabetNorm string) string {
	return strings.Map(func(r rune) rune {
		switch r {
		case 'A', 'E', 'I', 'O', 'U':
			return -1
		}
		return r
	}, alphabetNorm)
}
